### ISSUE TITLE
Assertion failure in vec_deque running rustc

### ISSUE THREAD

**Issue by thepowersgang** [thepowersgang/mrustc#142]
Using most recent HEAD (7ae47eff58abb672dd8462fda84bf652b348225b), built with `gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516`

```
[RUSTC] -o output/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore.rlib
thread 'main' panicked at 'assertion failed: ((self).head < (self).cap())', rustc-1.29.0-src/src/liballoc/collections/vec_deque.rs:360:47
```

**Comment by arlosi** [comment]
I hit this as well on Windows. I was able to successfully bootstrap rustc with [8e8d4fd6](https://github.com/thepowersgang/mrustc/commit/8e8d4fd6) + my 4 commits cherry-picked.

I never determined which commit(s) caused the regression. When debugging, it seemed like it might be related to 128 bit constant propagation.

**Comment by thepowersgang** [comment]
I think I've found the cause - it's a bad MIR optimisation that's lead to a borrow violation

The read of `_6*.1` was moved across the mutable borrow of `_6`'s source
```c
// <::"alloc"::collections::vec_deque::VecDeque<i32,>/*S*/ /*- <i32,>*/>::grow_if_necessary
...
        var6 = & arg0->_2;      // _6 = Borrow(0, Shared, a0*.2)
        var1 = & arg0->_2;      // _1 = Borrow(0, Unique, a0*.2)
        ZRING2c5alloc7raw_vec6RawVec2gCfNG2c5alloc5alloc6Global0g6double0g( var1 );
        // ^ Call( _2 = <::"alloc"::raw_vec::RawVec<i32,::"alloc"::alloc::Global/*S*/,>/*S*/ /*- <i32,::"alloc"::alloc::Global/*S*/,>*/>::double( _1, ), bb4, bb2)
        var3 = arg0;    // _3 = Borrow(0, Unique, a0*)
        ZRING3c5alloc11collections9vec_deque8VecDeque1gCf19handle_cap_increase0g( var3, var6->_1 );
        // ^ Call( _4 = <::"alloc"::collections::vec_deque::VecDeque<i32,>/*S*/ /*- <i32,>*/>::handle_cap_increase( _3, _6*.1, ), bb5, bb2)
```

**Comment by thepowersgang** [comment]
Fixed in d52931823e650451c22319980a8cc643b493a558 (with a separate fix related to 128-bit consteval fixing a different issue that broke bootstrap).

`rustc` now builds and runs successfully (able to build a runnable `hello`)
